lib/elixir_analyzer/test_suite/dna_encoding.ex

```diff
@@ -29,4 +30,127 @@ defmodule ElixirAnalyzer.TestSuite.DNAEncoding do
     called_fn name: :for
     comment Constants.dna_encoding_use_recursion()
   end
+
+  check_source "uses tail call recursion" do
+    type :essential
+    suppress_if "does not call any Enum functions", :fail
+    suppress_if "does not call any Stream functions", :fail
+    suppress_if "does not call any List functions", :fail
+    suppress_if "does not use list comprehensions", :fail
+    comment Constants.dna_encoding_use_tail_call_recursion()
+
+    check(%Source{code_ast: code_ast}) do
+      {_, tail_call_recursive_functions} =
+        Macro.prewalk(
+          code_ast,
+          [],
+          fn node, acc -> find_tail_call_recursive_functions(node, acc) end
+        )
+
+      tail_call_recursive_functions = Enum.uniq(tail_call_recursive_functions)
+
+      {_, all_recursive_functions} =
+        Macro.prewalk(
+          code_ast,
+          [],
+          fn node, acc -> find_all_recursive_functions(node, acc) end
+        )
+
+      all_recursive_functions = Enum.uniq(all_recursive_functions)
+
+      non_tail_call_recursive_functions = all_recursive_functions -- tail_call_recursive_functions
+
+      if non_tail_call_recursive_functions != [] || tail_call_recursive_functions == [] do
+        format_function_names = fn list ->
+          if list == [] do
+            "none"
+          else
+            Enum.map(list, fn {name, arity} -> "`#{name}/#{arity}`" end)
+            |> Enum.join(", ")
+          end
+        end
+
+        {false,
+         %{
+           non_tail_call_recursive_functions:
+             format_function_names.(non_tail_call_recursive_functions),
+           tail_call_recursive_functions: format_function_names.(tail_call_recursive_functions)
+         }}
+      else
+        true
+      end
+    end
+  end
+
+  defp find_tail_call_recursive_functions(node, acc) do
+    acc =
+      case node do
+        {op, _meta1, [{fn_name, _meta2, args}, opts]} when op in [:def, :defp] ->
```

Comment:
This is missing the `def  fn_name(args) when is_list(args) do ...` variant, right?

Reply:
Oops, fooled by guards again 😢 

Added support in [`17c59dd` (#374)](https://github.com/exercism/elixir-analyzer/pull/374/commits/17c59dd5e37747f7a323ccca165352bf76b8c274)